src/System.Threading.Thread/src/System/Threading/Thread.cs

```diff
@@ -0,0 +1,420 @@
+        public static long VolatileRead(ref long address) => Volatile.Read(ref address);
+        public static IntPtr VolatileRead(ref IntPtr address) => Volatile.Read(ref address);
+        public static object VolatileRead(ref object address) => Volatile.Read(ref address);
+        [CLSCompliant(false)]
+        public static sbyte VolatileRead(ref sbyte address) => Volatile.Read(ref address);
+        public static float VolatileRead(ref float address) => Volatile.Read(ref address);
+        [CLSCompliant(false)]
+        public static ushort VolatileRead(ref ushort address) => Volatile.Read(ref address);
+        [CLSCompliant(false)]
+        public static uint VolatileRead(ref uint address) => Volatile.Read(ref address);
+        [CLSCompliant(false)]
+        public static ulong VolatileRead(ref ulong address) => Volatile.Read(ref address);
+        [CLSCompliant(false)]
+        public static UIntPtr VolatileRead(ref UIntPtr address) => Volatile.Read(ref address);
+        public static void VolatileWrite(ref byte address, byte value) => Volatile.Write(ref address, value);
+        public static void VolatileWrite(ref double address, double value) => Volatile.Write(ref address, value);
+        public static void VolatileWrite(ref short address, short value) => Volatile.Write(ref address, value);
+        public static void VolatileWrite(ref int address, int value) => Volatile.Write(ref address, value);
+        public static void VolatileWrite(ref long address, long value) => Volatile.Write(ref address, value);
+        public static void VolatileWrite(ref IntPtr address, IntPtr value) => Volatile.Write(ref address, value);
+        public static void VolatileWrite(ref object address, object value) => Volatile.Write(ref address, value);
+        [CLSCompliant(false)]
+        public static void VolatileWrite(ref sbyte address, sbyte value) => Volatile.Write(ref address, value);
+        public static void VolatileWrite(ref float address, float value) => Volatile.Write(ref address, value);
+        [CLSCompliant(false)]
+        public static void VolatileWrite(ref ushort address, ushort value) => Volatile.Write(ref address, value);
+        [CLSCompliant(false)]
+        public static void VolatileWrite(ref uint address, uint value) => Volatile.Write(ref address, value);
+        [CLSCompliant(false)]
+        public static void VolatileWrite(ref ulong address, ulong value) => Volatile.Write(ref address, value);
+        [CLSCompliant(false)]
+        public static void VolatileWrite(ref UIntPtr address, UIntPtr value) => Volatile.Write(ref address, value);
+
+        /// <summary>
+        /// Manages functionality required to support members of <see cref="Thread"/> dealing with thread-local data
+        /// </summary>
+        private static class LocalDataStore
+        {
+            private static Dictionary<string, LocalDataStoreSlot> s_nameToSlotMap;
```

Comment:
There doesn't appear to be any mechanism provided here to ensure the thread-safety of the dictionary.